app/src/main/java/com/example/wellfed/ingredient/StoredIngredientDB.java

```diff
@@ -0,0 +1,307 @@
+                    }
+                });
+        // waiting for it to be finished
+        ingredientComplete.await();
+
+        // add it to stored ingredients now
+        Map<String, Object> stored = new HashMap<>();
+        stored.put("unit", storedIngredient.getUnit());
+        stored.put("amount", storedIngredient.getAmount());
+        stored.put("location", storedIngredient.getLocation());
+        stored.put("best-before", storedIngredient.getBestBefore());
+        DocumentReference documentReference = db.document("Ingredients/"+id);
+        stored.put("ingredient", documentReference);
+
+        CountDownLatch storageComplete = new CountDownLatch(1);
+        this.collection
+                .document(id)
+                .set(stored)
+                .addOnSuccessListener(new OnSuccessListener<Void>() {
+                    @Override
+                    public void onSuccess(Void aVoid) {
+                        Log.d(TAG, "DocumentSnapshot written with ID: " + id);
+                        storageComplete.countDown();
+                    }
+                })
+                .addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        Log.w(TAG, "Error adding document", e);
+                        storageComplete.countDown();
+                    }
+                });
+        storageComplete.await();
+        return id;
+    }
+
+    /**
+     * Updates a stored ingredient in the Firebase DB.
+     * @param id the ID of the ingredient to update
+     * @param storedIngredient the Ingredient containing the updated fields
+     * @throws InterruptedException when the on success listeners cannot complete
+     */
+    public void updateStoredIngredient(String id, StoredIngredient storedIngredient) throws InterruptedException {
+
+        Map<String, Object> ingredient = new HashMap<>();
+        ingredient.put("category", storedIngredient.getCategory());
+        ingredient.put("description", storedIngredient.getDescription());
+
+        CountDownLatch ingredientComplete = new CountDownLatch(1);
+        this.ingredients
+                .document(id)
+                .set(ingredient)
+                .addOnSuccessListener(new OnSuccessListener<Void>() {
+                    @Override
+                    public void onSuccess(Void aVoid) {
+                        Log.d(TAG, "DocumentSnapshot updated with ID: " + id);
+                        ingredientComplete.countDown();
+                    }
+                })
+                .addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        Log.w(TAG, "Error updating document", e);
+                        ingredientComplete.countDown();
+                    }
+                });
+
+        ingredientComplete.await();
```

Comment:
i suggest that we should use transactions to perform such queries. since a transaction guarantees all or nothing, cause this approach right now allows ingredient to be stored in the ingredient collection, but what if storing the storedingredient fails, now we have a lose ingredient.

Reply:
https://firebase.google.com/docs/firestore/manage-data/transactions

Reply:
I have added batch updates/adds/deletes to the document (which are all-or-nothing).